File: platform/hosted/gateway/tests/local/registry_runtime.py
#!/usr/bin/env python3
import argparse
import hashlib
import json
import os
from pathlib import Path
import secrets
import shlex
import signal
import stat
import struct
import subprocess
import sys


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--config", required=True)
    args = parser.parse_args()
    os.setsid()
    config = json.loads(Path(args.config).read_text())
    root = Path(config["root"]).resolve(strict=True)
    runtime = root / "registry-runtime"
    runtime.mkdir(mode=0o750)
    os.chown(runtime, 0, 4030)
    state = runtime / "state"
    state.mkdir(mode=0o700)
    os.chown(state, 4030, 4030)
    image = subprocess.check_output(
        ["docker", "image", "inspect", "--format", "{{.Id}}",
         "layerx-repair/registry-runtime:20260913"], text=True).strip()
    isolation_digest = subprocess.check_output(
        ["docker", "run", "--rm", image, "sha256sum", "/usr/bin/bwrap"],
        text=True).split()[0]
    bins = Path(config["service_bin_dir"]).resolve(strict=True)
    supervisor = bins / "layerx-cgroup-exec"
    supervisor_digest = hashlib.sha256(supervisor.read_bytes()).hexdigest()
    mounts = []
    mounted = {}

    def mount_file(source):
        source = Path(source)
        if source.resolve(strict=True) != source or not source.is_file():
            raise RuntimeError("registry input must be a canonical regular file")
        metadata = source.stat()
        if not stat.S_ISREG(metadata.st_mode) or metadata.st_nlink != 1:
            raise RuntimeError("registry input must be singly linked")
        if source not in mounted:
            os.chown(source, metadata.st_uid, 4030)
            os.chmod(source, 0o640)
            target = "/run/layerx/inputs/" + str(len(mounted))
            mounts.extend(["--mount", f"type=bind,src={source},dst={target},readonly"])
            mounted[source] = target
        return mounted[source]

    def generated(name, data):
        target = runtime / name
        descriptor = os.open(target, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
        with os.fdopen(descriptor, "wb") as output:
            output.write(data)
            output.flush()
            os.fsync(output.fileno())
        return mount_file(target)

    history = b"LayerX/sequencer-trust-history/v1\0"
    history += struct.pack(">HHHIQ", 1, 0, 3, config["network_id"], config["epoch"])
    history += bytes.fromhex(config["sequencer_id"])
    history += bytes.fromhex(config["sequencer_public_key"])
    history += struct.pack(">QQ", 1, (1 << 64) - 1) + bytes(9)
    certificates = Path(config["certificates_dir"])
    ca = mount_file(certificates / "ca.der")
    p12 = mount_file(config["client_pkcs12"])
    password = mount_file(config["client_password_file"])
    environment = {
        "LAYERX_REGISTRY_LISTEN": f"0.0.0.0:{config['listen_port']}",
        "LAYERX_REGISTRY_HOST_CGROUP_MOUNT": "/run/layerx/host-cgroup",
        "LAYERX_REGISTRY_STATE": "/run/layerx/state",
        "LAYERX_REGISTRY_BUILD_ROOT": "/run/layerx/quota",
        "LAYERX_REGISTRY_MAX_BUILDS": "1",
        "LAYERX_REGISTRY_BUILDER_IMAGE_DIGEST": Path(config["builder_digest_file"]).read_text().strip(),
        "LAYERX_REGISTRY_BUILDER_ENVIRONMENT_ROOT": "/run/layerx/builder",
        "LAYERX_REGISTRY_BUILDER_ENTRYPOINT": "/bin/layerx-build",
        "LAYERX_REGISTRY_BUILDER_ISOLATION_RUNTIME": "/usr/bin/bwrap",
        "LAYERX_REGISTRY_BUILDER_ISOLATION_RUNTIME_DIGEST": isolation_digest,
        "LAYERX_REGISTRY_BUILDER_JOB_SUPERVISOR": "/run/layerx/bin/layerx-cgroup-exec",
        "LAYERX_REGISTRY_BUILDER_JOB_SUPERVISOR_DIGEST": supervisor_digest,
        "LAYERX_REGISTRY_REQUEST_TOKEN_FILE": mount_file(config["request_token_file"]),
        "LAYERX_REGISTRY_PUBLICATION_TOKEN_FILE": generated("publication-token", secrets.token_hex(32).encode()),
        "LAYERX_REGISTRY_TLS_CERT_DER": mount_file(certificates / "core.der"),
        "LAYERX_REGISTRY_TLS_KEY_DER": mount_file(certificates / "core-key.der"),
        "LAYERX_REGISTRY_CLIENT_CA_DER": ca,
        "LAYERX_REGISTRY_OUTBOUND_CA_DER": ca,
        "LAYERX_REGISTRY_NODE_ENDPOINT": config["node_url"],
        "LAYERX_REGISTRY_RECEIPT_AUTHORITY_ENDPOINT": config["authority_url"],
        "LAYERX_REGISTRY_RECEIPT_AUTHORITY_REPLICA_ID": config["replica_id"],
        "LAYERX_REGISTRY_SEQUENCER_TRUST_HISTORY": generated("trust-history", history),
        "LAYERX_REGISTRY_IDENTITY_URL": config["identity_url"],
        "LAYERX_REGISTRY_IDENTITY_TOKEN_FILE": mount_file(config["identity_token_file"]),
        "LAYERX_REGISTRY_IDENTITY_CA_DER": ca,
        "LAYERX_REGISTRY_IDENTITY_CLIENT_IDENTITY_PKCS12": p12,
        "LAYERX_REGISTRY_IDENTITY_CLIENT_IDENTITY_PASSWORD_FILE": password,
    }
    for name, value in config["event_environment"].items():
        if name.startswith(("LAYERX_EVENTS_PROGRAM_", "LAYERX_EVENTS_WEBHOOKS_")):
            environment[name] = mount_file(value) if value.startswith("/") else value
    node_token = mount_file(config["node_token_file"])
    authority_token = mount_file(config["authority_token_file"])
    script = "#!/bin/sh\nset -eu\n"
    script += "\n".join(f"export {name}={shlex.quote(value)}" for name, value in environment.items()) + "\n"
    script += f'export LAYERX_REGISTRY_NODE_AUTHORIZATION="Bearer $(cat {shlex.quote(node_token)})"\n'
    script += f'export LAYERX_REGISTRY_RECEIPT_AUTHORITY_AUTHORIZATION="Bearer $(cat {shlex.quote(authority_token)})"\n'
    script += "exec /run/layerx/bin/layerx-program-registry\n"
    entrypoint = generated("entrypoint.sh", script.encode())
    name = f"layerx-registry-test-{os.getpid()}-{secrets.token_hex(4)}"
    command = ["docker", "run", "--rm", "--name", name, "--network", "host",
               "--cgroupns", "private", "--cap-drop", "ALL",
               "--cap-add", "CHOWN", "--cap-add", "SETUID", "--cap-add", "SETGID",
               "--security-opt", "no-new-privileges"]
    for source, target, readonly in [
        ("/sys/fs/cgroup", "/run/layerx/host-cgroup", False),
        (state, "/run/layerx/state", False),
        ("/root/lx-builds/repair-registry-runtime/quota", "/run/layerx/quota", False),
        (config["builder_root"], "/run/layerx/builder", True),
        (bins / "layerx-program-registry", "/run/layerx/bin/layerx-program-registry", True),
        (supervisor, "/run/layerx/bin/layerx-cgroup-exec", True),
    ]:
        command.extend(["--mount", f"type=bind,src={source},dst={target}" + (",readonly" if readonly else "")])
    command.extend(mounts)
    command.extend([image, "/bin/sh", entrypoint])
    process = None

    def stop(_signal, _frame):
        raise SystemExit(0)

    signal.signal(signal.SIGTERM, stop)
    signal.signal(signal.SIGINT, stop)
    try:
        process = subprocess.Popen(command)
        return process.wait()
    finally:
        subprocess.run(["docker", "stop", "--time", "5", name],
                       stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, check=False)
        if process is not None:
            process.wait(timeout=10)


if __name__ == "__main__":
    sys.exit(main())

File: platform/hosted/gateway/tests/local/registry_runtime.rs
use super::*;

pub fn configure(
    cluster: &Cluster,
    certificates: &Certificates,
    identity: &LocalIdentity,
    authority: &LocalAuthority,
    environment: &mut BTreeMap<&'static str, String>,
) -> (PathBuf, u16, Daemon) {
    let port = free_port();
    let request_token = local_secret(&cluster.root, "registry-request-token", &token());
    let node_port = free_port();
    let node_token = local_secret(&cluster.root, "registry-boundary-token", &token());
    let node = start_agent_boundary(cluster, certificates, node_port, &node_token);

    environment.insert(
        "LAYERX_GATEWAY_PROGRAM_REGISTRY_URL",
        format!("https://localhost:{port}"),
    );
    environment.insert(
        "LAYERX_GATEWAY_PROGRAM_REGISTRY_TOKEN_FILE",
        request_token.clone(),
    );
    let event_environment: BTreeMap<_, _> = environment
        .iter()
        .filter(|(name, _)| name.starts_with("LAYERX_EVENTS_"))
        .map(|(name, value)| (*name, value.clone()))
        .collect();
    let config = serde_json::json!({
        "root": cluster.root, "listen_port": port,
        "node_url": format!("https://localhost:{node_port}"),
        "node_token_file": node_token,
        "authority_url": format!("https://localhost:{}", authority.port),
        "authority_token_file": authority.token_file,
        "replica_id": hex_encode(&sha256(&[b"layerx-authority-replica:", hex_encode(&cluster.sequencer_key).as_bytes()])),
        "sequencer_id": hex_encode(&cluster.sequencer_id),
        "sequencer_public_key": hex_encode(&cluster.sequencer_key),
        "network_id": NETWORK_ID, "epoch": verified_epoch(cluster),
        "certificates_dir": certificates.path("ca.der").parent(),
        "service_bin_dir": std::env::var("LAYERX_TEST_REGISTRY_BIN_DIR").required("qualified registry binaries"),
        "runtime_image": std::env::var("LAYERX_TEST_REGISTRY_IMAGE").required("qualified registry runtime image"),
        "quota_root": std::env::var("LAYERX_TEST_REGISTRY_QUOTA_ROOT").required("qualified registry quota filesystem"),
        "builder_root": std::env::var("LAYERX_TEST_BUILDER_ROOT").required("qualified builder root"),
        "builder_digest_file": std::env::var("LAYERX_TEST_BUILDER_DIGEST_FILE").required("qualified builder digest"),
        "request_token_file": request_token,
        "event_environment": event_environment,
        "identity_url": format!("https://localhost:{}", identity.port),
        "identity_token_file": identity.tokens.join("registry"),
        "client_pkcs12": environment["LAYERX_GATEWAY_CLIENT_IDENTITY_PKCS12"],
        "client_password_file": environment["LAYERX_GATEWAY_CLIENT_IDENTITY_PASSWORD_FILE"],
        "lni_socket": cluster.lni_socket,
    });
    let path = cluster.root.join("registry-runtime.json");
    write(
        &path,
        serde_json::to_vec(&config)
            .required("registry runtime configuration")
            .as_slice(),
        0o600,
    );
    (path, port, node)
}

fn start_agent_boundary(
    cluster: &Cluster,
    certificates: &Certificates,
    port: u16,
    registry_token: &str,
) -> Daemon {
    let environment = BTreeMap::from([
        ("LAYERX_AGENT_BOUNDARY_LISTEN", format!("127.0.0.1:{port}")),
        (
            "LAYERX_AGENT_BOUNDARY_TLS_CERT_DER",
            text(&certificates.path("core.der")),
        ),
        (
            "LAYERX_AGENT_BOUNDARY_TLS_KEY_DER",
            text(&certificates.path("core-key.der")),
        ),
        (
            "LAYERX_AGENT_BOUNDARY_CLIENT_CA_DER",
            text(&certificates.path("ca.der")),
        ),
        (
            "LAYERX_AGENT_BOUNDARY_GATEWAY_TOKEN_FILE",
            local_secret(&cluster.root, "registry-boundary-gateway-token", &token()),
        ),
        (
            "LAYERX_AGENT_BOUNDARY_REGISTRY_TOKEN_FILE",
            registry_token.to_owned(),
        ),
        (
            "LAYERX_AGENT_BOUNDARY_WEBHOOK_TOKEN_FILE",
            local_secret(&cluster.root, "registry-boundary-webhook-token", &token()),
        ),
        (
            "LAYERX_AGENT_BOUNDARY_LNI_SOCKET",
            text(&cluster.lni_socket),
        ),
        (
            "LAYERX_AGENT_BOUNDARY_NODE_URL",
            format!("http://127.0.0.1:{}", cluster.program_port),
        ),
        (
            "LAYERX_AGENT_BOUNDARY_NODE_BEARER_TOKEN_FILE",
            local_secret(
                &cluster.root,
                "registry-native-node-token",
                &cluster.program_token,
            ),
        ),
        (
            "LAYERX_AGENT_BOUNDARY_STATE_DIR",
            text(&cluster.root.join("registry-node-boundary")),
        ),
        (
            "LAYERX_AGENT_BOUNDARY_PROTOCOL_NETWORK_ID",
            NETWORK_ID.to_string(),
        ),
        ("LAYERX_AGENT_BOUNDARY_NETWORK_ID", NETWORK_ID.to_string()),
    ]);
    local_service(cluster, "layerx-agent-boundary", port, &environment)
}

pub fn start(cluster: &Cluster, config: &Path, port: u16) -> Daemon {
    let python = std::env::var_os("LAYERX_TEST_PYTHON").required("qualified Python");
    let helper = repository_root().join("platform/hosted/gateway/tests/local/registry_runtime.py");
    let mut process = spawn(
        Path::new(&python),
        &[&text(&helper), "--config", &text(config)],
        &BTreeMap::new(),
        false,
        cluster.root.join("registry-runtime.stderr"),
    );
    process.supervised = true;
    wait_for_port(port, &mut process, "actual registry runtime");
    process
}

fn verified_epoch(cluster: &Cluster) -> u64 {
    let gate = ConnectionGate::new(1);
    let mut transport = must(
        Uds::connect(&cluster.lni_socket, &gate, lni_limits()),
        "registry LNI",
    );
    let handshake = must(
        perform(&mut transport, &handshake_config(), None),
        "registry handshake",
    );
    let header = must(
        layerx_client::batch::lookup(
            &mut transport,
            handshake.node().interface_version,
            handshake.node().latest_sealed_batch,
            81,
            cluster.sequencer_key,
        ),
        "registry signed batch",
    );
    must(
        layerx_wire::receipt::decode_batch_header(header.canonical_bytes()),
        "registry canonical header",
    )
    .epoch()
}
